### PR TITLE
REPL-only: Print a hint if the user types `exit` in the REPL

### PR DESCRIPTION
**Pull request by DilumAluthge**
Closes #21362
Closes #38307

This hint will only be shown if the input is `exit`.

```julia
julia> exit
exit (generic function with 2 methods)

To exit Julia, type exit() and press enter

julia> fs = [exit];

julia> fs[1]
exit (generic function with 2 methods)
```

### CODE REVIEW

**Comment by DilumAluthge** [comment]
@JeffBezanson Alright, I think I have finally figured out how to show the hint only when the REPL input is `exit`.

This does not affect how the `exit` object is shown.

File: stdlib/REPL/src/REPL.jl [hunk 1]
```diff
@@ -1239,4 +1242,15 @@ function start_repl_server(port::Int)
     end
 end
 
+print_exit_hint(repl::BasicREPL, line) = print_exit_hint(repl.terminal, line)
+print_exit_hint(repl::LineEditREPL, line) = print_exit_hint(repl.t, line)
+print_exit_hint(repl::StreamREPL, line) = print_exit_hint(repl.stream, line)
+function print_exit_hint(io::IO, line)
+    if lowercase(strip(line)) == "exit"
+        println(io)
+        println(io, "To exit Julia, type exit() and press enter")
```
Comment (simonbyrne) [inline stdlib/REPL/src/REPL.jl:1251]: Maybe mention better ways to exit?
```suggestion
        println(io, "To exit Julia use Ctrl-D, or type exit() and press enter")
```

Comment (simonbyrne) [reply stdlib/REPL/src/REPL.jl:1251]: I don't know if this works on Windows though.

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl]: We can do a Windows check, and only mention this if it isn't Windows.

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl]: Take a look now. We only mention `Ctrl-D` if it is not Windows.

File: stdlib/REPL/src/REPL.jl [hunk 2]
```diff
@@ -1239,4 +1242,20 @@ function start_repl_server(port::Int)
     end
 end
 
+print_exit_hint(repl::BasicREPL, line) = print_exit_hint(repl.terminal, line)
+print_exit_hint(repl::LineEditREPL, line) = print_exit_hint(repl.t, line)
+print_exit_hint(repl::StreamREPL, line) = print_exit_hint(repl.stream, line)
+function print_exit_hint(io::IO, line)
+    if lowercase(strip(line)) == "exit"
+        if Sys.iswindows()
+            msg = "To exit Julia, or type exit() and press enter"
```
Comment (simonbyrne) [inline stdlib/REPL/src/REPL.jl:1251]: Ctrl-D does seem to work on Windows, at least on the VM I tried?

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl:1251]: Cool. Let's just print it unconditionally.

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl:1251]: Well.... actually, is there a chance your VM software is remapping the keys for you?

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl:1251]: Several people on Slack also say that Control-D works to exit the Julia REPL on Windows.

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl]: Alright, we now mention Control-D unconditionally.

Comment (simonbyrne) [reply stdlib/REPL/src/REPL.jl]: Good to check, thanks!

File: stdlib/REPL/src/REPL.jl [hunk 3]
```diff
@@ -1239,4 +1242,14 @@ function start_repl_server(port::Int)
     end
 end
 
+print_exit_hint(repl::BasicREPL, line) = print_exit_hint(repl.terminal, line)
+print_exit_hint(repl::LineEditREPL, line) = print_exit_hint(repl.t, line)
+print_exit_hint(repl::StreamREPL, line) = print_exit_hint(repl.stream, line)
+function print_exit_hint(io::IO, line)
+    if lowercase(strip(line)) == "exit"
```
Comment (fingolfin) [inline stdlib/REPL/src/REPL.jl:1249]: Would it make sense to also print this if the input is "quit" (that's what e.g. GAP uses) ?

Anyway, I really like this PR, it fixes one of those tiny papercuts which seem irrelevant in isolation but add up.

**Review by fingolfin — APPROVED** [review]

File: stdlib/REPL/src/REPL.jl [hunk 4]
```diff
@@ -1239,4 +1242,14 @@ function start_repl_server(port::Int)
     end
 end
 
+print_exit_hint(repl::BasicREPL, line) = print_exit_hint(repl.terminal, line)
+print_exit_hint(repl::LineEditREPL, line) = print_exit_hint(repl.t, line)
+print_exit_hint(repl::StreamREPL, line) = print_exit_hint(repl.stream, line)
```
Comment (JeffBezanson) [inline stdlib/REPL/src/REPL.jl:1247]:
```
print_exit_hint(repl::AbstractREPL, line) = print_exit_hint(outstream(repl), line)
```

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl]: Done!

File: stdlib/REPL/src/REPL.jl (hunk 3 above)
Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl:1247]: I wouldn't object to also doing this for "quit", but I'll defer to @JeffBezanson.

I'd want to keep this limited to a very small set of inputs. For example, you can exit R by typing `q()` and pressing enter. But I definitely wouldn't want to print a hint if someone types `q` in the Julia REPL.

I think it makes sense for `exit` because the corresponding function `exit()` is a function that actually exists in Julia. But we don't actually have a `quit` function in Julia.

Comment (DilumAluthge) [reply stdlib/REPL/src/REPL.jl:1247]: But I don't feel strongly either way. Happy to add `quit`, also happy to leave it out.

File: stdlib/REPL/src/REPL.jl [hunk 5]
```diff
@@ -1239,4 +1242,12 @@ function start_repl_server(port::Int)
     end
 end
 
+print_exit_hint(repl::AbstractREPL, line) = print_exit_hint(outstream(repl), line)
+function print_exit_hint(io::IO, line)
+    if lowercase(strip(line)) == "exit"
+        println(io, "\nTo exit Julia, use Ctrl-D, or type exit() and press enter")
```
Comment (JeffBezanson) [inline stdlib/REPL/src/REPL.jl:1248]:
```suggestion
        println(io, "\nTo exit Julia, use Ctrl-D, or type exit() and press enter.")
```

File: stdlib/REPL/src/REPL.jl (hunk 3 above)
Comment (JeffBezanson) [reply stdlib/REPL/src/REPL.jl:1247]: Yes I think we should add it.

Comment (JeffBezanson) [reply stdlib/REPL/src/REPL.jl:1247]: I like what the python repl does:
```
>>> quit
Use quit() or Ctrl-D (i.e. EOF) to exit
>>> quit = 2
>>> quit
2
```
It stops printing the message if `quit` is defined as a variable. Stylish! I think we should do this too.
Can be checked with `Base.isbindingresolved(Main, :exit) && isdefined(Main, :exit)`.

File: doc/build/build.md [hunk 6]
```diff
@@ -276,3 +276,15 @@ distribution does not include the source code of dependencies.
 
 For example, `julia-1.0.0.tar.gz` is the light source distribution for the `v1.0.0` release
 of Julia, while `julia-1.0.0-full.tar.gz` is the full source distribution.
+
+## Building Julia from source with a Git checkout of one or more stdlibs
```
Comment (fingolfin) [inline doc/build/build.md:280]: Accident?

Comment (DilumAluthge) [reply doc/build/build.md:280]: Ugh. Yeah I pushed to the wrong branch.

Comment (DilumAluthge) [reply doc/build/build.md:280]: I should be able to force-push to the previous commit.

Comment (DilumAluthge) [reply doc/build/build.md]: Should be fixed now.

File: stdlib/REPL/src/REPL.jl (hunk 3 above)
Comment (vtjnash) [reply stdlib/REPL/src/REPL.jl:1248]: In python, that's not magic, it is just literally the definition of show for that default variable:
```
>>> exit.__class__
<class 'site.Quitter'>
>>> quit.__class__
<class 'site.Quitter'>
```

The equivalent here would be:
```
julia> using Base: exit as quit; # @eval Base const quit = exit; @eval Base export quit;

julia> Base.show(io::IO, ::MIME"text/plain", ::typeof(exit)) = print(io, "Use exit() or Ctrl-D (i.e. EOF) to quit")

julia> repr(exit)
"exit"

julia> show(exit)
exit
julia> summary(exit)
"typeof(exit)"

julia> display(exit)
Use exit() or Ctrl-D (i.e. EOF) to quit

julia> exit
Use exit() or Ctrl-D (i.e. EOF) to quit

julia> exit()
```

Comment (vtjnash) [reply stdlib/REPL/src/REPL.jl:1248]: Which would have been https://github.com/JuliaLang/julia/pull/21362

File: stdlib/REPL/src/REPL.jl [hunk 7]
```diff
@@ -1232,4 +1235,20 @@ function run_frontend(repl::StreamREPL, backend::REPLBackendRef)
...
 end
 
+function start_repl_server(port::Int)
+    return listen(port) do server, status
+        client = accept(server)
+        run_repl(client)
+        nothing
+    end
+end
+
```
Comment (vtjnash) [inline stdlib/REPL/src/REPL.jl:1245]:
```suggestion
```

**Comment by mkitti** [comment]
@DilumAluthge asked if I would be interested in making the final push on this. Here's are a few thoughts:

1. `quit` should be handled in exactly the same manner as `help` in #41754. Perhaps that should be a separate PR entirely.
2. For `exit`, we're adding an additional parsing step for every command entered into the REPL by hard coding in a call to `print_exit_hint` which then calls `lowercase(strip(line))`.
3. As alternative, we could just add a default `ast_transform` and we can match the abstract syntax tree rather than trying to parse the line again. If someone doesn't like this, they could just `pop!(Base.active_repl_backend.ast_transforms)`.

An example `ast_transform` might work like this:
```julia
julia> function print_exit_hint(ast)
           if !isempty(ast.args) && ast.args[end] == :exit
               quote
                   println("To exit Julia, use Ctrl-D, or type exit() and press enter.")
                   exit
               end
           else
               ast
           end
       end
print_exit_hint (generic function with 1 method)

julia> push!(Base.active_repl_backend.ast_transforms, print_exit_hint)
2-element Vector{Any}:
 softscope (generic function with 1 method)
 print_exit_hint (generic function with 1 method)

julia> exit
To exit Julia, use Ctrl-D, or type exit() and press enter.
exit (generic function with 2 methods)

julia> out = ans
exit (generic function with 2 methods)

julia> typeof(out)
typeof(exit)

julia> pop!(Base.active_repl_backend.ast_transforms)
print_exit_hint (generic function with 1 method)

julia> exit
exit (generic function with 2 methods)
```

**Comment by mkitti** [comment]
The hint for `quit` is over at #41990 which addresses @JeffBezanson 's https://github.com/JuliaLang/julia/pull/38522#discussion_r533695883 .

If we wanted to go the ast transform route, I've prepared this commit on my branch of  `dpa/exit-hint-repl`
https://github.com/mkitti/julia/commit/6bb3909100e82b81fcedc55510c8151591b56ac4

That version is stricter than the version above and tries to parallel the `softscope` ast transform in form.
```julia
julia> Base.isbindingresolved(Main, :exit)
true

julia> isdefined(Main, :exit)
true

julia> exit = 2
ERROR: cannot assign a value to variable Base.exit from module Main
Stacktrace:
 [1] top-level scope
   @ REPL[10]:1

julia> exit
exit (generic function with 2 methods)

julia> function print_exit_hint(@nospecialize ex)
           if ex isa Expr && ex.head === :toplevel && length(ex.args) == 2 && ex.args[2] === :exit
               quote
                   println("suggestion: To exit Julia, use Ctrl-D, or type exit() and press enter.")
                   exit
               end
           else
               return ex
           end
       end
print_exit_hint (generic function with 1 method)

julia> push!(Base.active_repl_backend.ast_transforms, print_exit_hint)
2-element Vector{Any}:
 softscope (generic function with 1 method)
 print_exit_hint (generic function with 1 method)

julia> exit
suggestion: To exit Julia, use Ctrl-D, or type exit() and press enter.
exit (generic function with 2 methods)

julia> repr(exit)
"exit"

julia> summary(exit)
"exit (generic function with 2 methods)"

julia> display(exit)
exit (generic function with 2 methods)

julia> filter!(!=(print_exit_hint), Base.active_repl_backend.ast_transforms)
1-element Vector{Any}:
 softscope (generic function with 1 method)

julia> exit
exit (generic function with 2 methods)

julia> exit()
```

I also suggest adding the prefix "suggestion: " to be consistent with the other REPL hints.

Also note above that we cannot assign a new value to `exit`, so I am unsure what @JeffBezanson meant in the last sentence of https://github.com/JuliaLang/julia/pull/38522#discussion_r533695883 .

I would appreciate if reviewers could address these questions:
1. Do we stick with the original implementation or refactor to an abstract syntax tree transform?
2. If we do use the AST transform, is the order of the lines and the whitespace acceptable?
3. Are there any other unaddressed issues?

**Comment by DilumAluthge** [comment]
> Also note above that we cannot assign a new value to `exit`, so I am unsure what @JeffBezanson meant in the last sentence of [#38522 (comment)](https://github.com/JuliaLang/julia/pull/38522#discussion_r533695883) .

You can't assign a new value to `exit` if the binding has already been resolved:
```julia
julia> exit
exit (generic function with 2 methods)

julia> exit = 123
ERROR: cannot assign a value to variable Base.exit from module Main
Stacktrace:
 [1] top-level scope
   @ REPL[1]:1
```

But you can assign a new value to `exit` if the binding has not yet been resolved. For example, run the following in a brand-new REPL session:
```julia
julia> exit = 123
123

julia> exit
123

julia> exit()
ERROR: MethodError: objects of type Int64 are not callable
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```

**Comment by DilumAluthge** [comment]
So I think you'd need to check something like `Main.exit === Base.exit` to figure out if the user has redefined `exit` to something else.

**Comment by mkitti** [comment]
> So I think you'd need to check something like `Main.exit === Base.exit` to figure out if the user has redefined `exit` to something else.

How does one redefine exit in Main?

```
julia> exit = 2
ERROR: cannot assign a value to variable Base.exit from module Main
Stacktrace: [1] top-level scope @ REPL[10]:1
```

**Comment by DilumAluthge** [comment]
> > So I think you'd need to check something like `Main.exit === Base.exit` to figure out if the user has redefined `exit` to something else.
> 
> How does one redefine exit in Main?
> 
> ```
> julia> exit = 2
> ERROR: cannot assign a value to variable Base.exit from module Main
> Stacktrace: [1] top-level scope @ REPL[10]:1
> ```

Huh. If I run `exit = 2` in a brand-new REPL session in Julia 1.6.2, it works fine for me. 


What Julia version did you try it in?

Also, make sure it's a brand-new session.

**Comment by DilumAluthge** [comment]
```julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.6.2 (2021-07-14)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> exit = 123
123

julia> exit
123

julia> exit()
ERROR: MethodError: objects of type Int64 are not callable
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```

```julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.7.0-beta2 (2021-06-20)
 _/ |\__'_|_|_|\__'_|  |  Official https://julialang.org/ release
|__/                   |

julia> exit = 456
456

julia> exit
456

julia> exit()
ERROR: MethodError: objects of type Int64 are not callable
Maybe you forgot to use an operator such as *, ^, %, / etc. ?
Stacktrace:
 [1] top-level scope
   @ REPL[3]:1
```
